app/lib/differential_expression_service.rb

```diff
@@ -1,5 +1,116 @@
+    raise ArgumentError, "#{study.accession} has no default annotation" if study.default_annotation.blank?
+
+    annotation_name, annotation_type, annotation_scope = study.default_annotation.split('--')
+    annotation = {
+      annotation_name: annotation_name,
+      annotation_type: annotation_type,
+      annotation_scope: annotation_scope
+    }
+    cluster_file = study.default_cluster.study_file
+    requested_user = user || study.user
+    run_differential_expression_job(cluster_file, study, requested_user, **annotation)
+  end
+
+  # same as above method, except runs differential expression job on all valid annotations
+  #
+  # * *params*
+  #   - +study_accession+ (String) => Accession of study to use
+  #   - +user+ (User) => Corresponding user, will default to study owner
+  #
+  # * *yields*
+  #   - (IngestJob) => Differential expression job in PAPI for each valid cluster/annotation combination
+  #
+  # * *returns*
+  #   - (Integer) => Number of DE jobs yielded
+  #
+  # * *raises*
+  #   - (ArgumentError) => if requested study cannot run any DE jobs
+  def self.run_differential_expression_on_all(study_accession, user: nil)
+    study = Study.find_by(accession: study_accession)
+    validate_study(study)
+    available_annotations = []
+
+    metadata = study.cell_metadata.where(annotation_type: 'group', is_differential_expression_enabled: false)
+                    .select(&:can_visualize?)
+    available_annotations += metadata.map do |meta|
+      { annotation_name: meta.name, annotation_type: meta.annotation_type, annotation_scope: 'study' }
+    end
+
+    cell_annotations = []
+    study.cluster_groups.each do |cluster|
+      next if cluster.cell_annotations.empty?
+
+      cell_annots = cluster.cell_annotations.select do |annot|
+        annot['type'] == 'group' &&
+          cluster.can_visualize_cell_annotation?(annot) &&
+          !annot[:is_differential_expression_enabled]
+      end
+
+      cell_annots.each do |annot|
+        annot[:cluster_file_id] = cluster.study_file.id # for checking associations later
+      end
+
+      cell_annotations += cell_annots
+    end
+
+    available_annotations += cell_annotations.map do |annot|
+      {
+        annotation_name: annot[:name],
+        annotation_type: annot[:type],
+        annotation_scope: 'cluster',
+        cluster_file_id: annot[:cluster_file_id]
+      }
+    end
+    raise ArgumentError, "#{study_accession} does not have any valid annotations" if available_annotations.empty?
+
+    log_message "#{study_accession} has data available for DE; validating inputs"
+    requested_user = user || study.user
+
+    job_count = 0
+    study.cluster_ordinations_files.each do |cluster_file|
+      available_annotations.each do |annotation|
+        begin
+          # skip if this is a cluster-based annotation and is not available on this cluster file
+          next if annotation[:scope] == 'cluster' && annotation[:cluster_file_id] != cluster_file.id
```

Comment:
same as above, I think this will be clearer and more debuggable if the filter is done outside the `each`

Reply:
This one is a little trickier as it requires the `each` since we're comparing the annotation against the current clustering file.  I don't think the above approach will work.